diff --git a/src/ducc0/fft/fft1d_impl.h b/src/ducc0/fft/fft1d_impl.h
index eb698375..93a30aef 100644
--- a/src/ducc0/fft/fft1d_impl.h
+++ b/src/ducc0/fft/fft1d_impl.h
@@ -86,13 +86,7 @@ using namespace std;
 namespace {
 
 template<typename T> constexpr inline size_t fft1d_simdlen
-  = min<size_t>(8, native_simd<T>::size());
-template<> constexpr inline size_t fft1d_simdlen<double>
-  = min<size_t>(4, native_simd<double>::size());
-template<> constexpr inline size_t fft1d_simdlen<float>
-  = min<size_t>(8, native_simd<float>::size());
-template<typename T> using fft1d_simd = typename simd_select<T,fft1d_simdlen<T>>::type;
-template<typename T> constexpr inline bool fft1d_simd_exists = (fft1d_simdlen<T> > 1);
+  = native_simd<T>::size();
 
 template<bool fwd, typename T> void ROTX90(Cmplx<T> &a)
   { auto tmp_= fwd ? -a.r : a.r; a.r = fwd ? a.i : -a.i; a.i=tmp_; }
@@ -1257,9 +1251,9 @@ template <typename Tfs> class cfft_multipass: public cfftpass<Tfs>
         }
       else
         {
-        if constexpr(is_same<T,Tfs>::value && fft1d_simd_exists<Tfs>) // we can vectorize!
+        if constexpr(is_same<T,Tfs>::value && simd_exists<Tfs, native_simd<Tfs>::size()>) // we can vectorize!
           {
-          using Tfv = fft1d_simd<Tfs>;
+          using Tfv = native_simd<Tfs>;
           using Tcv = Cmplx<Tfv>;
           constexpr size_t vlen = Tfv::size();
           size_t nvtrans = (l1*ido + vlen-1)/vlen;
diff --git a/src/ducc0/fft/fftnd_impl.h b/src/ducc0/fft/fftnd_impl.h
index 50effc95..b3d983a9 100644
--- a/src/ducc0/fft/fftnd_impl.h
+++ b/src/ducc0/fft/fftnd_impl.h
@@ -95,15 +95,6 @@ using namespace std;
 
 namespace {
 
-template<typename T> constexpr inline size_t fft_simdlen
-  = min<size_t>(8, native_simd<T>::size());
-template<> constexpr inline size_t fft_simdlen<double>
-  = min<size_t>(4, native_simd<double>::size());
-template<> constexpr inline size_t fft_simdlen<float>
-  = min<size_t>(8, native_simd<float>::size());
-template<typename T> using fft_simd = typename simd_select<T,fft_simdlen<T>>::type;
-template<typename T> constexpr inline bool fft_simd_exists = (fft_simdlen<T> > 1);
-
 struct util // hack to avoid duplicate symbols
   {
   static void sanity_check_axes(size_t ndim, const shape_t &axes)
@@ -165,12 +156,12 @@ struct util // hack to avoid duplicate symbols
     }
 
   static size_t thread_count (size_t nthreads, const fmav_info &info,
-    size_t axis, size_t /*vlen*/)
+    size_t axis, size_t vlen)
     {
     if (nthreads==1) return 1;
     size_t size = info.size();
-    if (size<32768) return 1;  // not worth opening a parallel region
-    size_t max_parallel = size / info.shape(axis);
+    if (size<=32768) return 1;  // not worth opening a parallel region
+    size_t max_parallel = size / (info.shape(axis)*vlen);
     size_t max_threads = adjust_nthreads(nthreads);
     return std::max(size_t(1), std::min(max_parallel, max_threads));
     }
@@ -397,7 +388,7 @@ template<typename T, typename T0> class TmpStorage
         d.realloc(bufsize_trafo);
         return;
         }
-      constexpr auto vlen = fft_simdlen<T0>;
+      constexpr auto vlen = native_simd<T0>::size();
       // FIXME: when switching to C++20, use bit_floor(othersize)
       size_t buffct = std::min(vlen, n_trafo);
       size_t datafct = std::min(vlen, n_trafo);
@@ -405,8 +396,8 @@ template<typename T, typename T0> class TmpStorage
       dstride = bufsize_data;
       dofs = bufsize_trafo;
       // critical stride avoidance
-      if ((dstride&256)==0) dstride+=16;
-      if ((dofs&256)==0) dofs += 16;
+      if ((dstride!=0)&&((dstride&256)==0)) dstride+=16;
+      if ((dofs!=0)&&((dofs&256)==0)) dofs += 16;
       d.realloc(buffct*dofs + datafct*dstride);
       }
 
@@ -439,13 +430,11 @@ template <typename Tsimd, typename Titer> DUCC0_NOINLINE void copy_input(const T
   const Cmplx<typename Tsimd::value_type> * DUCC0_RESTRICT ptr = src.data();
   for (size_t i=0; i<it.length_in(); ++i)
     {
-    Cmplx<Tsimd> tmp;
     for (size_t j=0; j<vlen; ++j)
       {
-      tmp.r[j] = ptr[it.iofs(j,i)].r;
-      tmp.i[j] = ptr[it.iofs(j,i)].i;
+      dst[i].r[j] = ptr[it.iofs(j,i)].r;
+      dst[i].i[j] = ptr[it.iofs(j,i)].i;
       }
-    dst[i] = tmp;
     }
   }
 
@@ -456,10 +445,8 @@ template <typename Tsimd, typename Titer> DUCC0_NOINLINE void copy_input(const T
   const typename Tsimd::value_type * DUCC0_RESTRICT ptr = src.data();
   for (size_t i=0; i<it.length_in(); ++i)
     {
-    typename Tsimd::value_type tmp[vlen];
     for (size_t j=0; j<vlen; ++j)
-      tmp[j] = ptr[it.iofs(j,i)];
-    dst[i] = loadu<Tsimd>(&tmp[0]);
+      dst[i][j] = ptr[it.iofs(j,i)];
     }
   }
 
@@ -479,9 +466,8 @@ template<typename Tsimd, typename Titer> DUCC0_NOINLINE void copy_output(const T
   Cmplx<typename Tsimd::value_type> * DUCC0_RESTRICT ptr = dst.data();
   for (size_t i=0; i<it.length_out(); ++i)
     {
-    Cmplx<Tsimd> tmp(src[i]);
     for (size_t j=0; j<vlen; ++j)
-      ptr[it.oofs(j,i)].Set(tmp.r[j],tmp.i[j]);
+      ptr[it.oofs(j,i)].Set(src[i].r[j],src[i].i[j]);
     }
   }
 
@@ -492,9 +478,8 @@ template<typename Tsimd, typename Titer> DUCC0_NOINLINE void copy_output(const T
   typename Tsimd::value_type * DUCC0_RESTRICT ptr = dst.data();
   for (size_t i=0; i<it.length_out(); ++i)
     {
-    Tsimd tmp = src[i];
     for (size_t j=0; j<vlen; ++j)
-      ptr[it.oofs(j,i)] = tmp[j];
+      ptr[it.oofs(j,i)] = src[i][j];
     }
   }
 
@@ -514,15 +499,11 @@ template <typename Tsimd, typename Titer> DUCC0_NOINLINE void copy_input(const T
   for (size_t i=0; i<it.length_in(); ++i)
     for (size_t j0=0; j0<nvec; ++j0)
       {
-      typename Tsimd::value_type tmp[2*vlen];
       for (size_t j1=0; j1<vlen; ++j1)
         {
-        tmp[j1] = ptr[it.iofs(j0*vlen+j1,i)].r;
-        tmp[j1+vlen] = ptr[it.iofs(j0*vlen+j1,i)].i;
+        dst[j0*vstr+i].r[j1] = ptr[it.iofs(j0*vlen+j1,i)].r;
+        dst[j0*vstr+i].i[j1] = ptr[it.iofs(j0*vlen+j1,i)].i;
         }
-
-      dst[j0*vstr+i].r = loadu<Tsimd>(&tmp[0]);
-      dst[j0*vstr+i].i = loadu<Tsimd>(&tmp[vlen]);
       }
   }
 template <typename T, typename Titer> DUCC0_NOINLINE void copy_input(const Titer &it,
@@ -542,10 +523,8 @@ template <typename Tsimd, typename Titer> DUCC0_NOINLINE void copy_input(const T
   for (size_t i=0; i<it.length_in(); ++i)
     for (size_t j0=0; j0<nvec; ++j0)
       {
-      typename Tsimd::value_type tmp[vlen];
       for (size_t j1=0; j1<vlen; ++j1)
-        tmp[j1] = ptr[it.iofs(j0*vlen+j1,i)];
-      dst[j0*vstr+i] = loadu<Tsimd>(&tmp[0]);
+        dst[j0*vstr+i][j1] = ptr[it.iofs(j0*vlen+j1,i)];
       }
   }
 
@@ -566,9 +545,8 @@ template<typename Tsimd, typename Titer> DUCC0_NOINLINE void copy_output(const T
   for (size_t i=0; i<it.length_out(); ++i)
     for (size_t j0=0; j0<nvec; ++j0)
       {
-      Cmplx<Tsimd> tmp(src[j0*vstr+i]);
       for (size_t j1=0; j1<vlen; ++j1)
-        ptr[it.oofs(j0*vlen+j1,i)].Set(tmp.r[j1],tmp.i[j1]);
+        ptr[it.oofs(j0*vlen+j1,i)].Set(src[j0*vstr+i].r[j1],src[j0*vstr+i].i[j1]);
       }
   }
 template<typename T, typename Titer> DUCC0_NOINLINE void copy_output(const Titer &it,
@@ -587,9 +565,8 @@ template<typename Tsimd, typename Titer> DUCC0_NOINLINE void copy_output(const T
   for (size_t i=0; i<it.length_out(); ++i)
     for (size_t j0=0; j0<nvec; ++j0)
       {
-      Tsimd tmp(src[j0*vstr+i]);
       for (size_t j1=0; j1<vlen; ++j1)
-        ptr[it.oofs(j0*vlen+j1,i)] = tmp[j1];
+        ptr[it.oofs(j0*vlen+j1,i)] = src[j0*vstr+i][j1];
       }
   }
 template<typename T, typename Titer> DUCC0_NOINLINE void copy_output(const Titer &it,
@@ -632,10 +609,10 @@ DUCC0_NOINLINE void general_nd(const cfmav<T> &in, const vfmav<T> &out,
         plan : get_plan<Tplan>(len, true);
       }
 
-    execParallel(util::thread_count(nthreads, in, axes[iax], fft_simdlen<T0>),
+    execParallel(util::thread_count(nthreads, in, axes[iax], native_simd<T0>::size()),
       [&](Scheduler &sched)
       {
-      constexpr auto vlen = fft_simdlen<T0>;
+      constexpr auto vlen = native_simd<T0>::size();
       constexpr size_t nmax = 16;
       const auto &tin(iax==0? in : out);
       multi_iter<nmax> it(tin, out, axes[iax], sched.num_threads(), sched.thread_num());
@@ -671,9 +648,15 @@ DUCC0_NOINLINE void general_nd(const cfmav<T> &in, const vfmav<T> &out,
         }
       else  // fairly small individual FFT, vectorizing probably beneficial
         {
-        // if no stride, only vectorize if vectorized FFT fits into cache
+        // if no stride, only vectorize up to a total size that still fits into L2
         // if strided, always vectorize (TBC)
-        n_simul = nostride ? ((wss(vlen)<=l2cache) ? vlen:1) : vlen;
+        if (nostride)
+          {
+          n_simul = vlen;
+          while((n_simul>1) && (wss(n_simul)>l2cache)) n_simul /= 2;
+          }
+        else
+          n_simul = vlen;
         if (critstride)  // make bunch large to reduce overall copy cost
           {
           n_bunch=n_simul;
@@ -1035,9 +1018,9 @@ template<typename T> DUCC0_NOINLINE void general_r2c(
   auto plan = get_plan<pocketfft_r<T>>(in.shape(axis), in.ndim()==1);
   size_t len=in.shape(axis);
   execParallel(
-    util::thread_count(nthreads, in, axis, fft_simdlen<T>),
+    util::thread_count(nthreads, in, axis, native_simd<T>::size()),
     [&](Scheduler &sched) {
-    constexpr auto vlen = fft_simdlen<T>;
+    constexpr auto vlen = native_simd<T>::size();
     TmpStorage<T,T> storage(in.size()/len, len, plan->bufsize(), 1, false);
     multi_iter<vlen> it(in, out, axis, sched.num_threads(), sched.thread_num());
 #ifndef DUCC0_NO_SIMD
@@ -1153,9 +1136,9 @@ template<typename T> DUCC0_NOINLINE void general_c2r(
   auto plan = get_plan<pocketfft_r<T>>(out.shape(axis), in.ndim()==1);
   size_t len=out.shape(axis);
   execParallel(
-    util::thread_count(nthreads, in, axis, fft_simdlen<T>),
+    util::thread_count(nthreads, in, axis, native_simd<T>::size()),
     [&](Scheduler &sched) {
-      constexpr auto vlen = fft_simdlen<T>;
+      constexpr auto vlen = native_simd<T>::size();
       TmpStorage<T,T> storage(out.size()/len, len, plan->bufsize(), 1, false);
       multi_iter<vlen> it(in, out, axis, sched.num_threads(), sched.thread_num());
 #ifndef DUCC0_NO_SIMD
@@ -1684,9 +1667,9 @@ template<typename Tplan, typename T0, typename T, typename Exec>
   plan1->exec(fkernel.data(), T0(1)/T0(l_in), true, nthreads);
 
   execParallel(
-    util::thread_count(nthreads, in, axis, fft_simdlen<T0>),
+    util::thread_count(nthreads, in, axis, native_simd<T0>::size()),
     [&](Scheduler &sched) {
-      constexpr auto vlen = fft_simdlen<T0>;
+      constexpr auto vlen = native_simd<T0>::size();
       TmpStorage<T,T0> storage(in.size()/l_in, l_in+l_out, bufsz, 1, false);
       multi_iter<vlen> it(in, out, axis, sched.num_threads(), sched.thread_num());
 #ifndef DUCC0_NO_SIMD
